src/decorators/OrmRepository.ts

```diff
@@ -1,21 +1,179 @@
+ * ```ts
+ * class Sample {
+ *   // constructor injection
+ *   constructor(
+ *     \@OrmRepository("test-conn")
+ *      private userRepository: UserRepository,
+ *   ) {}
+ * 
+ *   // property injection
+ *  \@OrmRepository("test-conn")
+ *   userRepository: UserRepository;
+ * }
+ * ```
+ */
+export function OrmRepository(connectionName: string): ParamOrPropDecorator;
+/**
+ * Allows to inject a Repository, MongoRepository, TreeRepository using TypeDI's Container
+ * and specify the connection name in a parameter.
+ * Be aware that you have to annotate the param/property with correct type!
+ * ```ts
+ * class Sample {
+ *   // constructor injection
+ *   constructor(
+ *     \@OrmRepository(User, "test-conn")
+ *      private userRepository: Repository<User>,
+ *   ) {}
+ * 
+ *   // property injection
+ *  \@OrmRepository(User, "test-conn")
+ *   userRepository: Repository<User>;
+ * }
+ * ```
+ */
+export function OrmRepository(entityType: Function, connectionName: string): ParamOrPropDecorator;
+
+export function OrmRepository(entityTypeOrConnectionName?: Function|string, paramConnectionName = "default"): ParamOrPropDecorator {
+    return (object: object, propertyName: string, index?: number) => {
+        let entityType: Function|undefined;
+        let connectionName: string;
+        let repositoryType: Function;
+
+        // handle first parameter overload
+        connectionName = paramConnectionName;
+        if (typeof entityTypeOrConnectionName === "string") {
+            connectionName = entityTypeOrConnectionName;
+        } else if (typeof entityTypeOrConnectionName === "function") {
+            entityType = entityTypeOrConnectionName;
+        }
+
+        // if the decorator has been aplied to parameter (constructor injection)
+        if (index) {
+            const paramTypes: Function[] | undefined = Reflect.getOwnMetadata("design:paramtypes", object, propertyName);
+            if (!paramTypes || !paramTypes[index]) {
+                throw new Error(
+                    `Cannot get reflected type for a "${propertyName}" method's ${index + 1}. parameter of ${object.constructor.name} class. ` +
+                    `Make sure you have turned on an "emitDecoratorMetadata": true, option in tsconfig.json. ` +
+                    `and that you have imported "reflect-metadata" on top of the main entry file in your application.` +
+                    `And make sure that you have annotated the property type correctly with: ` +
+                    `Repository, MongoRepository, TreeRepository or custom repository class type.`
+                );
+            }
+            repositoryType = paramTypes[index];
+        }
+        // if the parameter has been aplied to class property
+        else {
+            const propertyType: Function | undefined = Reflect.getOwnMetadata("design:type", object, propertyName);
+            if (!propertyType) {
+                throw new Error(
```

Comment:
maybe better to extract errors in to "error" classes? It makes code so huge and hardly readable

Reply:
Right, I haven't thought that the decorator code will grow up to this amount of lines 😉